lib/experimental/lexopt.nim

```diff
@@ -0,0 +1,251 @@
+  l.cmdline[l.index]
+
+proc isLongOpt*(s: string): bool = s.startsWith("--")
+  ## Returns whether `s` can be parsed as a long option.
+proc isShortOpt*(s: string): bool = s.startsWith('-')
+  ## Returns whether `s` can be parsed as a short option.
+proc isDash(s: string): bool = s == "-"
+proc isDashDash(s: string): bool = s == "--"
+
+proc prefix*(kind: CmdlineKind): string =
+  ## Return the prefix string for a given command line parameter kind
+  case kind
+  of cmdLong:
+    "--"
+  of cmdShort:
+    "-"
+  else:
+    ""
+
+func newUnexpectedValueError(opt, value: string): ref UnexpectedValueError =
+  (ref UnexpectedValueError)(
+    msg: "unexpected value '" & value & "' for '" & opt & "'",
+    opt: opt,
+    value: value
+  )
+
+proc next*(l: var CmdLexer): (CmdlineKind, string) =
+  ## Consume the next option.
+  ##
+  ## The following section describes how various syntaxes are handled:
+  ##
+  ## 1. `--` and `-` are considered values.
+  ## 2. `--foo:bar` and `--foo=bar` are treated as long options with key `foo`
+  ##    and `bar` is considered the value of `foo`.
+  ##
+  ##    If this value is not consumed before the next call to `next()`, an error
+  ##    will be raised. See below for more details.
+  ## 3. `--foo` is treated as a long option with key `foo`.
+  ## 4. `-abco` is treated as into `a`, `b`, `c`, `o` short options, returned
+  ##    one per `next()` call.
+  ## 5. `-abco:foo` and `-abco=foo` are treated as `a`, `b`, `c`, `o` short
+  ##    options, with `foo` being the value of option `o`.
+  ##
+  ##    If this value is not consumed before the next call to `next()`, an error
+  ##    will be raised. See below for more details
+  ## 6. Everything else is considered values.
+  ##
+  ## If the previous option have an unconsumed value (eg. `value()` was not
+  ## called for `--opt:foo`), `UnexpectedValueError` will be raised.
+  type
+    LexState {.pure.} = enum
+      ## Current lexer state
+      Value ## A value that is not an option
+      ShortOpt ## A short option
+      LongOpt ## A long option
+      LongValue ## Value of a long option
+      ShortValue ## Value of a short option
+      Ended ## Nothing left
+
+  let currentState =
+    if l.index >= l.cmdline.len:
+      LexState.Ended
+    elif l.current.isLongOpt:
+      if l.valueIdx > 0:
+        LexState.LongValue
+      elif not l.current.isDashDash:
+        LexState.LongOpt
+      else:
+        LexState.Value
+    elif l.current.isShortOpt:
+      if l.valueIdx > 0 and l.current[l.valueIdx] in ValueDelims:
+        LexState.ShortValue
+      elif not l.current.isDash:
+        LexState.ShortOpt
+      else:
+        LexState.Value
+    else:
+      LexState.Value
+
+  case currentState
+  of LongOpt:
+    let delimIdx = l.current.find(ValueDelims)
+    if delimIdx > 2:
+      l.valueIdx = delimIdx
+      result = (cmdLong, l.current[2..<delimIdx])
+    else:
+      result = (cmdLong, l.current[2..^1])
+      inc l.index
```

Comment:
Not a strong opinion, but I personally think disallowing `--:` and `--=` would be better. If they stay allowed, I think there should be comment here mentioning that this is intentional.

Reply:
I'm not big in introducing the concept of syntax errors for command line, where no rigid grammar exists, which is why these cases are in a bit of a limbo.

I decided to have them parse into something that can be caught by simple code, for example:

```nim
let (kind, opt) = lexer.next()
case kind
of cmdLong:
  case opt
  of "foo":
  of "bar":
  else:
    # error handling would handle `--=something` automatically
```

I can't say that this is the best way to handle it, but I think keeping this as an "implementation-defined" case is not a bad idea, hence the lack of any comments confirming whether this is intended or not.

Reply:
The case being implementation-defined and subject to change is fine, in my opinion, but due the lack of comment, I initially thought that this was a case you hadn't considered. Only after looking at the tests did it became clear that the behaviour is intended.

With comment, I didn't mean that it needs to be mentioned in the doc comment, just that there should be a comment above the marked code saying something along the lines of "`--:` not being rejected is deliberate", so that future readers don't trip on it like I did.

Reply:
A comment was added to the other branch instead.